src/client/auth/oidc.rs

```diff
@@ -23,44 +24,117 @@ use crate::{
 use super::{sasl::SaslContinue, Credential, MONGODB_OIDC_STR};
 
 /// The user-supplied callbacks for OIDC authentication.
+pub struct State {
+    callback: Callback,
+    cache: Arc<RwLock<Cache>>,
+}
+
+impl Clone for State {
+    fn clone(&self) -> Self {
+        Self {
+            callback: self.callback.clone(),
+            cache: Arc::new(RwLock::new(Cache::default())),
+        }
+    }
+}
+
+impl State {
+    pub(crate) async fn get_refresh_token(&self) -> Option<String> {
+        self.cache.read().await.refresh_token.clone()
+    }
+
+    // TODO RUST-1662: This function will actually be used.
+    #[allow(dead_code)]
+    pub(crate) async fn get_access_token(&self) -> Option<String> {
+        self.cache.read().await.access_token.clone()
+    }
+}
+
+// TODO RUST-1497: This enum will be public
+#[allow(dead_code)]
 #[derive(Clone)]
-pub struct Callbacks {
-    inner: Arc<CallbacksInner>,
+enum Callback {
```

Comment:
Since the contained value here is `Arc<CallbackInner>` in both cases, this could also be
```
struct Callback {
  inner: Arc<CallbackInner>,
  kind: CallbackKind,
}
enum CallbackKind {
  Machine,
  Human,
}
```

Which to use is largely a matter of taste and convenience; IMO, if the expectation is that inner state will diverge, or if the code handling them is mostly disjoint, the current way is fine, but if it's mostly common handling I'd rather have the `kind` tag.

Reply:
This is actually something I was actually considering, too, so I'm happy to see you bring it up. I think this idea makes the code a little easier to read, and it really has no affect on users of the library since they will always act through the `human` and `machine` methods.